File: tools/zed-vize/assert-zed-package.mjs
#!/usr/bin/env node
import assert from "node:assert/strict";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import zlib from "node:zlib";

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../..");
const archivePath = path.resolve(
  process.cwd(),
  process.argv[2] ?? path.join(root, "zed-vize-extension.tar.gz"),
);

assert.ok(fs.existsSync(archivePath), `Zed extension archive does not exist: ${archivePath}`);

const archiveSize = fs.statSync(archivePath).size;
assert.ok(archiveSize > 5_000, `Zed extension archive is suspiciously small: ${archiveSize} bytes`);
assert.ok(
  archiveSize < 2_000_000,
  `Zed extension archive is unexpectedly large: ${archiveSize} bytes`,
);

const entries = readTarGz(archivePath);
const entryNames = entries.map((entry) => entry.name).sort((a, b) => a.localeCompare(b));
const entryMap = new Map(entries.map((entry) => [entry.name, entry]));

assert.deepEqual(
  entryNames,
  Array.from(new Set(entryNames)),
  "Zed archive contains duplicate entries",
);

for (const name of entryNames) {
  assert.ok(!name.includes("\\"), `Zed archive entry must use POSIX separators: ${name}`);
  assert.ok(!name.includes("\0"), `Zed archive entry contains a NUL byte: ${name}`);
  assert.ok(!name.startsWith("/"), `Zed archive entry must be relative: ${name}`);
  assert.ok(!name.split("/").includes(".."), `Zed archive entry must not traverse: ${name}`);
  assert.ok(name === "zed/" || name.startsWith("zed/"), `unexpected root: ${name}`);
}

const requiredFiles = [
  "zed/Cargo.lock",
  "zed/Cargo.toml",
  "zed/LICENSE",
  "zed/README.md",
  "zed/extension.toml",
  "zed/languages/art-vue/brackets.scm",
  "zed/languages/art-vue/config.toml",
  "zed/languages/art-vue/highlights.scm",
  "zed/languages/art-vue/indents.scm",
  "zed/languages/art-vue/injections.scm",
  "zed/languages/art-vue/outline.scm",
  "zed/languages/art-vue/overrides.scm",
  "zed/src/lib.rs",
];

for (const name of requiredFiles) {
  assert.ok(entryMap.has(name), `Zed archive is missing required file: ${name}`);
  assert.ok(readTextEntry(entryMap, name).trim().length > 0, `Zed archive file is empty: ${name}`);
}

const allowedEntries = [
  /^zed\/$/,
  /^zed\/Cargo\.lock$/,
  /^zed\/Cargo\.toml$/,
  /^zed\/LICENSE$/,
  /^zed\/README\.md$/,
  /^zed\/extension\.toml$/,
  /^zed\/languages\/$/,
  /^zed\/languages\/art-vue\/$/,
  /^zed\/languages\/art-vue\/(?:brackets|config|highlights|indents|injections|outline|overrides)\.scm$/,
  /^zed\/languages\/art-vue\/config\.toml$/,
  /^zed\/src\/$/,
  /^zed\/src\/lib\.rs$/,
];

for (const name of entryNames) {
  assert.ok(
    allowedEntries.some((pattern) => pattern.test(name)),
    `Zed archive ships an unexpected file: ${name}`,
  );
}

const forbiddenEntries = [
  /^zed\/\.git/,
  /^zed\/\.github\//,
  /^zed\/\.zed\//,
  /^zed\/node_modules\//,
  /^zed\/target\//,
  /\/\.DS_Store$/,
  /~$/,
];

for (const name of entryNames) {
  for (const pattern of forbiddenEntries) {
    assert.ok(!pattern.test(name), `Zed archive must not ship ${name}`);
  }
}

const workspaceVersion = readWorkspaceVersion();
const extensionToml = readTextEntry(entryMap, "zed/extension.toml");
const cargoToml = readTextEntry(entryMap, "zed/Cargo.toml");
const libRs = readTextEntry(entryMap, "zed/src/lib.rs");
const artVueConfig = readTextEntry(entryMap, "zed/languages/art-vue/config.toml");
const injections = readTextEntry(entryMap, "zed/languages/art-vue/injections.scm");
const highlights = readTextEntry(entryMap, "zed/languages/art-vue/highlights.scm");

assertTomlString(extensionToml, "id", "vize");
assertTomlString(extensionToml, "name", "Vize");
assertTomlString(extensionToml, "version", workspaceVersion);
assertTomlString(extensionToml, "repository", "https://github.com/ubugeeei-prod/vize");
assert.match(extensionToml, /^\[language_servers\.vize\]$/m);
assert.match(extensionToml, /^languages = \["Vue", "Art Vue"\]$/m);
assert.match(extensionToml, /^\[language_servers\.vize\.language_ids\]$/m);
assert.match(extensionToml, /^"Vue" = "vue"$/m);
assert.match(extensionToml, /^"Art Vue" = "art-vue"$/m);
assert.match(extensionToml, /^\[grammars\.art-vue\]$/m);
assert.match(extensionToml, /^commit = "[0-9a-f]{40}"$/m);

assertTomlString(cargoToml, "name", "vize-zed-extension");
assertTomlString(cargoToml, "version", workspaceVersion);
assertTomlString(cargoToml, "edition", "2024");
assertTomlString(cargoToml, "license", "MIT");
assert.match(cargoToml, /^publish = false$/m);
assert.match(cargoToml, /^crate-type = \["cdylib"\]$/m);
assert.match(cargoToml, /^zed_extension_api = "=0\.7\.0"$/m);

assert.match(libRs, /const SERVER_NAME: &'static str = "vize";/);
assert.match(libRs, /const SERVER_BINARY: &'static str = "vize";/);
assert.match(libRs, /worktree\.which\(Self::SERVER_BINARY\)/);
assert.match(libRs, /unwrap_or_else\(\|\| vec!\["lsp"\.to_string\(\)\]\)/);
assert.match(libRs, /language_server_initialization_options/);
assert.match(libRs, /language_server_workspace_configuration/);
assert.match(libRs, /zed::register_extension!\(VizeExtension\);/);

assertTomlString(artVueConfig, "name", "Art Vue");
assertTomlString(artVueConfig, "grammar", "art-vue");
assert.match(artVueConfig, /^path_suffixes = \["art\.vue"\]$/m);
assertTomlString(artVueConfig, "prettier_parser_name", "vue");

for (const queryFile of [
  "brackets",
  "highlights",
  "indents",
  "injections",
  "outline",
  "overrides",
]) {
  const contents = readTextEntry(entryMap, `zed/languages/art-vue/${queryFile}.scm`);
  assert.ok(contents.trim().length > 0, `${queryFile}.scm must not be empty`);
}

assert.match(injections, /directive_attribute/);
assert.match(injections, /style_element/);
assert.match(injections, /template_element/);
assert.match(highlights, /@tag\.component\.type\.constructor/);

console.log(
  `Zed package smoke passed: ${path.relative(root, archivePath)} (${entryNames.length} entries)`,
);

function assertTomlString(source, key, expected) {
  assert.match(source, new RegExp(`^${escapeRegExp(key)} = "${escapeRegExp(expected)}"$`, "m"));
}

function readWorkspaceVersion() {
  const cargoToml = fs.readFileSync(path.join(root, "Cargo.toml"), "utf-8");
  const version = cargoToml.match(/^version = "(.+)"$/m)?.[1];
  assert.ok(version, "workspace version is missing from Cargo.toml");
  return version;
}

function readTextEntry(entryMap, name) {
  const entry = entryMap.get(name);
  assert.ok(entry, `missing tar entry: ${name}`);
  return entry.data.toString("utf-8");
}

function readTarGz(filePath) {
  const buffer = zlib.gunzipSync(fs.readFileSync(filePath));
  const entries = [];
  let offset = 0;

  while (offset + 512 <= buffer.byteLength) {
    const header = buffer.subarray(offset, offset + 512);
    if (header.every((byte) => byte === 0)) {
      break;
    }

    const name = readTarString(header, 0, 100);
    const prefix = readTarString(header, 345, 155);
    const fullName = prefix ? `${prefix}/${name}` : name;
    const size = parseOctal(readTarString(header, 124, 12));
    const typeflag = readTarString(header, 156, 1) || "0";
    const dataOffset = offset + 512;
    const data = buffer.subarray(dataOffset, dataOffset + size);

    assert.ok(fullName, "tar entry name must not be empty");
    if (typeflag === "x" || typeflag === "g") {
      offset = dataOffset + Math.ceil(size / 512) * 512;
      continue;
    }

    assert.ok(
      typeflag === "0" || typeflag === "5",
      `unsupported tar entry type ${typeflag}: ${fullName}`,
    );

    entries.push({ data, name: fullName, size, typeflag });
    offset = dataOffset + Math.ceil(size / 512) * 512;
  }

  return entries;
}

function readTarString(buffer, offset, length) {
  const raw = buffer.subarray(offset, offset + length);
  const end = raw.indexOf(0);
  return raw.subarray(0, end === -1 ? raw.byteLength : end).toString("utf-8");
}

function parseOctal(value) {
  const trimmed = value.trim();
  return trimmed ? Number.parseInt(trimmed, 8) : 0;
}

function escapeRegExp(value) {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

File: tools/commands/editors/zed/assert-zed-package.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [package]
//! edition = "2024"
//!
//! [dependencies]
//! serde = { version = "1", features = ["derive"] }
//! serde_json = "1"
//! ```

use std::{collections::BTreeSet, env, path::PathBuf, process::ExitCode};

#[path = "../../../rust/common.rs"]
mod common;
#[path = "../../../rust/editor_archive.rs"]
mod editor_archive;

fn main() -> ExitCode {
    common::main_result(run())
}

fn run() -> Result<(), String> {
    let root = common::repo_root()?;
    let archive = archive_path(&root, "zed-vize-extension.tar.gz");
    editor_archive::assert_size(&archive, "Zed extension archive", 5_000, 2_000_000)?;
    let mut entries = editor_archive::list_tar_gz(&archive)?;
    entries.sort();
    editor_archive::assert_unique(&entries, "Zed archive")?;
    editor_archive::assert_safe_entries(&entries, "Zed archive", "zed")?;
    let required = [
        "zed/Cargo.lock",
        "zed/Cargo.toml",
        "zed/LICENSE",
        "zed/README.md",
        "zed/extension.toml",
        "zed/languages/art-vue/brackets.scm",
        "zed/languages/art-vue/config.toml",
        "zed/languages/art-vue/highlights.scm",
        "zed/languages/art-vue/indents.scm",
        "zed/languages/art-vue/injections.scm",
        "zed/languages/art-vue/outline.scm",
        "zed/languages/art-vue/overrides.scm",
        "zed/src/lib.rs",
    ];
    editor_archive::require_entries(
        &archive,
        &entries,
        &required,
        editor_archive::read_tar_text,
        "Zed archive",
    )?;
    assert_allowed(
        &entries,
        &[
            "zed/",
            "zed/Cargo.lock",
            "zed/Cargo.toml",
            "zed/LICENSE",
            "zed/README.md",
            "zed/extension.toml",
            "zed/languages/",
            "zed/languages/art-vue/",
            "zed/languages/art-vue/brackets.scm",
            "zed/languages/art-vue/config.toml",
            "zed/languages/art-vue/highlights.scm",
            "zed/languages/art-vue/indents.scm",
            "zed/languages/art-vue/injections.scm",
            "zed/languages/art-vue/outline.scm",
            "zed/languages/art-vue/overrides.scm",
            "zed/src/",
            "zed/src/lib.rs",
        ],
        "Zed archive",
    )?;
    assert_no_forbidden(
        &entries,
        "Zed archive",
        &[
            ".git",
            ".github/",
            ".zed/",
            "node_modules/",
            "target/",
            ".DS_Store",
            "~",
        ],
    )?;
    let workspace_version = editor_archive::workspace_version(&root)?;
    let extension = editor_archive::read_tar_text(&archive, "zed/extension.toml")?;
    let cargo = editor_archive::read_tar_text(&archive, "zed/Cargo.toml")?;
    let lib = editor_archive::read_tar_text(&archive, "zed/src/lib.rs")?;
    let art_config = editor_archive::read_tar_text(&archive, "zed/languages/art-vue/config.toml")?;
    let injections =
        editor_archive::read_tar_text(&archive, "zed/languages/art-vue/injections.scm")?;
    let highlights =
        editor_archive::read_tar_text(&archive, "zed/languages/art-vue/highlights.scm")?;
    for (source, needle) in [
        (&extension, r#"id = "vize""#),
        (&extension, r#"name = "Vize""#),
        (&extension, &format!(r#"version = "{workspace_version}""#)),
        (
            &extension,
            r#"repository = "https://github.com/ubugeeei-prod/vize""#,
        ),
        (&extension, "[language_servers.vize]"),
        (&extension, r#"languages = ["Vue", "Art Vue"]"#),
        (&extension, "[language_servers.vize.language_ids]"),
        (&extension, r#""Vue" = "vue""#),
        (&extension, r#""Art Vue" = "art-vue""#),
        (&extension, "[grammars.vue]"),
        (&cargo, r#"name = "vize-zed-extension""#),
        (&cargo, &format!(r#"version = "{workspace_version}""#)),
        (&cargo, r#"edition = "2024""#),
        (&cargo, r#"license = "MIT""#),
        (&cargo, "publish = false"),
        (&cargo, r#"crate-type = ["cdylib"]"#),
        (&cargo, r#"zed_extension_api = "=0.7.0""#),
        (&lib, r#"const SERVER_NAME: &'static str = "vize";"#),
        (&lib, r#"const SERVER_BINARY: &'static str = "vize";"#),
        (&lib, "worktree.which(Self::SERVER_BINARY)"),
        (&lib, r#"unwrap_or_else(|| vec!["lsp".to_string()])"#),
        (&lib, "language_server_initialization_options"),
        (&lib, "recommended_initialization_options"),
        (&lib, r#""editor": true"#),
        (&lib, r#""ecosystem": true"#),
        (&lib, r#""lint": true"#),
        (&lib, r#""typecheck": true"#),
        (&lib, "language_server_workspace_configuration"),
        (&lib, "zed::register_extension!(VizeExtension);"),
        (&art_config, r#"name = "Art Vue""#),
        (&art_config, r#"grammar = "vue""#),
        (&art_config, r#"path_suffixes = ["art.vue"]"#),
        (&art_config, r#"prettier_parser_name = "vue""#),
        (&injections, "directive_attribute"),
        (&injections, "style_element"),
        (&injections, "template_element"),
        (&highlights, "@tag.component.type.constructor"),
    ] {
        editor_archive::expect_contains(source, needle, &format!("Zed package missing {needle}"))?;
    }
    if extension.contains("[grammars.") && extension.contains("[grammars.art-vue]") {
        return Err("Zed extension.toml must not declare dashed grammar ids".to_string());
    }
    println!(
        "Zed package smoke passed: {} ({} entries)",
        common::relative_path(&root, &archive),
        entries.len()
    );
    Ok(())
}

fn archive_path(root: &std::path::Path, default: &str) -> PathBuf {
    env::args_os()
        .nth(1)
        .map(PathBuf::from)
        .unwrap_or_else(|| root.join(default))
}

fn assert_allowed(entries: &[String], allowed: &[&str], label: &str) -> Result<(), String> {
    let allowed = allowed.iter().copied().collect::<BTreeSet<_>>();
    for entry in entries {
        if !allowed.contains(entry.as_str()) {
            return Err(format!("{label} ships an unexpected file: {entry}"));
        }
    }
    Ok(())
}

fn assert_no_forbidden(entries: &[String], label: &str, forbidden: &[&str]) -> Result<(), String> {
    for entry in entries {
        for pattern in forbidden {
            if entry.contains(pattern) || entry.ends_with(pattern) {
                return Err(format!("{label} must not ship {entry}"));
            }
        }
    }
    Ok(())
}
